Fix: Return early response for already onboarded users sending onboarding messages

N1kunj1998 opened the pull request:
## Summary
Prevents onboarding messages from already onboarded users from being processed by LLM/vector store. Returns immediate "already registered" response instead of IDK.

## Changes
- Skip LLM processing for onboarding messages in `process.py`
- Return custom "already registered" message in `generate.py` 
- Use static `RELATED_QUESTIONS` instead of dynamic fetching
- Enhanced `is_onboard()` for case-insensitive matching across languages
- Added unit and integration tests

## Testing
Run integration test to verify:
python byoeb-v1/byoeb/tests/integration/test_early_return_already_onboarded.py
- `process.py` - Skip LLM for onboarding messages
- `generate.py` - Early return with custom response
- `onboarding_text.py` - Added `ALREADY_REGISTERED_DICT`
- `utils.py` - Improved onboarding detection
- `whatsapp.py` - Fixed message type handling
- Added tests in `tests/unit/` and `tests/integration/`

mamuqsit committed:
fastmcp: pin version to 2.14.0 (#98)

fastmcp.Client in v2.14.1 does not work - it hangs indefinitely

mamuqsit committed:
ci: rename staging branch (#97)

* ci: rename staging branch

* docs/deployment.md: update branch name

N1kunj1998 committed:
feat: Add WhatsApp template messages for onboarding, leaderboard, and DYK with multi-language support (#95)

* background-jobs: deprecate POST /schedule endpoint (#93)

* background-jobs: deprecate POST /schedule endpoint

* Return 200 OK with JSON "detail" as per standards
Addresses few points made by @a4i-architect

* POST /asha_logs: tweak setup, add integration tests (#92)

* POST /asha_logs: add integration tests
test_admin.py is expected to fail at two points

* admin.py: fix POST /asha_logs not returning expected response
This change is expected to fix the failing integration tests

* test_admin: use datetime.isoformat() in place of custom date formatter

* /asha_logs: stream logs server-side, process file client-side
Unlike XLSX, CSV is one of the better formats to use here as we can iteratively deliver data to user.

XLSX is a ZIP of XML files so it cannot be streamed.

* /get_bot_messages: add a 'limit=1000' parameter
We can leverage this to get a sample of messages for later /asha_logs integration tests

* test_admin: add test case to assert /asha_logs csv output format
This test will ensure the format contains few necessary attributes and that all records are sorted by incoming timestamp attribute.
/get_bot_messages is used to fetch a sample of up to a 100 messages.

* cleanup code, diff

* workflows: temporarily run cd
This change will be reverted after failing test is fixed

* Revert "workflows: temporarily run cd"

This reverts commit dbff763d4879516d839948bf142c17351613fa09.

* changed leaderboard and send dyk code to use template instade of sending normal whatsapp message

* added template message for user to get message when he is onboarded via backed script

* final commit also checked by running background jobs and made changes to now work with template code

* Quote all DYK CSV fields to avoid comma parsing errors

* Cleanup DYK templates and enforce CSV validation

---------

Co-authored-by: Muqsit <notmuqsit+a4i@gmail.com>

mamuqsit committed:
tests: cleanup, widen onboarding test to cover all languages (#100)

* refactor: cleanup onboarding test, extend to all four supported languages

* test-onboarding: validate new user attributes

* test-onboarding: cleanup requests

* test-onboarding: reformat code

* .gitignore: add coverage.json

* byoeb-integrations: cleanup tests - avoid creating persistent .wav files

* tests: patch for sqlite v3.35

N1kunj1998 committed:
Automated Monthly ASHABot Log Analysis (#82)

* inital commit for this task, will remove background jobs change in future commit

* removed background jobs part and removed unnecessary code and created new script for extracting relavent data

* reuse asha_logs.fetch_daily_logs in compile_monthly_logs script

* added code for IDK analysis

* Added more data points and results of IDK analysis in output markdown file

* background-jobs: deprecate POST /schedule endpoint (#93)

* background-jobs: deprecate POST /schedule endpoint

* Return 200 OK with JSON "detail" as per standards
Addresses few points made by @a4i-architect

* added LLM generated insights and summary

* POST /asha_logs: tweak setup, add integration tests (#92)

* POST /asha_logs: add integration tests
test_admin.py is expected to fail at two points

* admin.py: fix POST /asha_logs not returning expected response
This change is expected to fix the failing integration tests

* test_admin: use datetime.isoformat() in place of custom date formatter

* /asha_logs: stream logs server-side, process file client-side
Unlike XLSX, CSV is one of the better formats to use here as we can iteratively deliver data to user.

XLSX is a ZIP of XML files so it cannot be streamed.

* /get_bot_messages: add a 'limit=1000' parameter
We can leverage this to get a sample of messages for later /asha_logs integration tests

* test_admin: add test case to assert /asha_logs csv output format
This test will ensure the format contains few necessary attributes and that all records are sorted by incoming timestamp attribute.
/get_bot_messages is used to fetch a sample of up to a 100 messages.

* cleanup code, diff

* workflows: temporarily run cd
This change will be reverted after failing test is fixed

* Revert "workflows: temporarily run cd"

This reverts commit dbff763d4879516d839948bf142c17351613fa09.

* added charts and Key Metrics table and also code for MoM analysis

* Completed code with adding upload to Azure functionality

* added user related analysis and MoM growth and changed upload container in azure storage

* Use: Handle monthly log fetch from async generator before building DataFrame

---------

Co-authored-by: Muqsit <notmuqsit+a4i@gmail.com>

N1kunj1998 committed:
task completed and includes integration tests as well for testing

a4i-architect reviewed (commented):
PR Review Summary:  

The code introduces important enhancements aimed at robust handling of onboarding and message processing, especially in multilingual contexts and for the "already onboarded" early exit logic. Additionally, the integration test script provides comprehensive validation of these features. Below is a detailed feedback:

**Highlights & Strengths:**  
- Clear handling of "already registered" messages per language, avoiding unintended reprocessing.  
- Use of static dictionary `RELATED_QUESTIONS` mitigates dependency issues, ensuring consistent related questions.  
- Enhanced debug logging for message flow and cache operations improves observability.  
- The test script covers multiple languages, onboarding message variants, and normal flows, which promotes reliability.  
- Database operations now use `_id` for update robustness, reducing potential errors with nested filters.  
- Validation report generation with detailed per-language validation enhances post-test insights.

**Concerns & Areas for Improvement:**  

1. **Code Complexity & Readability:**  
   - **Deep nesting and duplicated logic** (e.g., in `is_onboard`) makes maintenance harder. Simplify normalization and matching logic, possibly by abstracting common string processing tasks into helper functions. (See [Effective Python: Writing clear and maintainable code](https://effectivepython.com/).)  
   - The `__create_user_message` function contains convoluted logic around `message_type` inference, with multiple conditionals that could be streamlined.  
   
2. **Cyclomatic Complexity & Duplication:**  
   - The logic determining `message_type` in `__create_user_message` and in message handling could be refactored into dedicated helper functions to reduce complexity.  
   - The validation function repeats similar checks per language, which can be centralized with better data structures, reducing code duplication.

3. **Type Safety & Robustness:**  
   - The database update logic relies on `_id`, which is good; but fallback mechanisms should explicitly log fallbacks for clarity.  
   - In `utils.is_onboard`, currently, the string normalization (e.g., replacing hyphens, whitespace) is partial; more consistent normalization (e.g., using regex, case normalization carefully per language) would reduce false negatives.

4. **Logging & Traceability:**  
   - While debug prints are helpful during development, they should be wrapped in a debug flag or replaced with proper logging. As-is, excessive print statements may clutter logs and impact performance.

5. **Test Code Review:**  
   - The test script is comprehensive but is quite monolithic. Modularize into small functions per step (send, wait, log, validate) for better granularity.  
   - Multiple sleep calls (`asyncio.sleep`) are used to wait for processing; consider using event-based synchronization if possible.

6. **Handling of Translations & Edge Cases:**  
   - The `text_translator.atranslate_text` call assumes always successful translation; error handling should be added.  
   - Falling back to source text for cases where translation fails (e.g., empty response) should be considered.

7. **Design & Patterns:**  
   - Overall, the code adopts an imperative, step-by-step logic which is understandable but can benefit from more modularization and pattern use (e.g., state or command pattern for message handling).  
   - The validation and test reporting are quite verbose—consider encapsulating into classes for potential extension.

**Typings & Style:**  
- Some functions lack type hints; adding them consistently would enhance readability.  
- Consistent use of f-strings with careful truncation (`[:100]`) is good, but placeholders like `None` should be handled cleanly to avoid errors.

**Conclusion & Score (out of 100):**  
**75/100**  
The code introduces solid features but suffers from over-complication, some duplicate logic, and missing robustness checks. It is functional but can be refined for clarity, maintainability, and performance.

**Relevant Articles/Web References:**  
- [Python’s Best Practices - Clean Code](https://realpython.com/clean-code-python/)  
- [Refactoring for Readability in Python](https://refactoring.guru/refactoring)  
- [Effective Python: 90 Specific Ways to Write Better Python](https://effectivepython.com/)  
- [Design Patterns for Python](https://refactoring.guru/design-patterns/python)

---

**Summary:**  
- Good step forward with detailed logging and robust DB updates.  
- Needs refactoring of complex conditional logic, normalization functions, and test modularization.  
- Suggest moving print debugging to a proper logging framework, and adding error handling where assumptions are made (e.g., translation).  
- Overall, a decent implementation with room for maintainability improvements.

Acceptance threshold met with **75**/100.

mamuqsit reviewed (approved):
Tested with MCP, functionality is working - ASHABot explicitly states the user is already onboarded.

```jsonc
// user_language=en
{
  "category": "bot_to_asha_response",
  "text": "You are already registered with the system. You can ask me any question about pregnancy, childcare, ...",
  "additional_info": [
    ["Internal query", "onboard asha"],
    [
      "Related questions",
      [
        "How much does a 1-year-old typically weigh?",
        "What long-term effects does tobacco cause?",
        "What is Antara injection?"
      ]
    ]
  ]
}

// user_language=hi
{
  "category": "bot_to_asha_response",
  "text": "आप पहले से ही सिस्टम में पंजीकृत हैं। आप गर्भावस्था, शिशु देखभाल, ... के बारे में मुझसे कोई भी सवाल पूछ सकते हैं।",
  "additional_info": [
    [
      "Internal query",
      "onboard-asha"
    ],
    [
      "संबंधित सवाल",
      [
        "1 साल का बच्चा आमतौर पर कितना वज़न रखता है?",
        "तंबाकू के दीर्घकालिक प्रभाव क्या होते हैं?",
        "अंतरा इंजेक्शन क्या है?"
      ]
    ]
  ]
}

// user_language=mr
{
  "category": "bot_to_asha_response",
  "text": "तुम्ही आधीच सिस्टममध्ये नोंदणीकृत आहात. तुम्ही गर्भावस्था, शिशु देखभाल, ... याबद्दल मला कोणताही प्रश्न विचारू शकता.",
  "additional_info": [
    [
      "Internal query",
      "onboard-asha"
    ],
    [
      "संबंधित प्रश्न",
      [
        "1 वर्षाचा मुलगा सामान्यतः किती वजनाचा असतो?",
        "तंबाकूचे दीर्घकालीन परिणाम काय आहेत?",
        "अंतरा इंजेक्शन म्हणजे काय?"
      ]
    ]
  ]
}

// user_language=te
{
  "category": "bot_to_asha_response",
  "text": "మీరు ఇప్పటికే సిస్టమ్‌లో నమోదు చేయబడ్డారు. మీరు గర్భధారణ, శిశు సంరక్షణ, ... గురించి నన్ను ఏదైనా ప్రశ్న అడగవచ్చు.",
  "additional_info": [
    [
      "Internal query",
      "onboard-asha"
    ],
    [
      "సంబంధిత ప్రశ్నలు",
      [
        "ఒక సంవత్సరానికి చెందిన బిడ్డ సాధారణంగా ఎంత బరువు ఉంటుంది?",
        "తంబాకుకు దీర్ఘకాలిక ప్రభావాలు ఏమిటి?",
        "అంతర ఇంజెక్షన్ అంటే ఏమిటి?"
      ]
    ]
  ]
}
```

mamuqsit commented on byoeb-v1/byoeb/tests/integration/test_early_return_already_onboarded.py:
```diff
@@ -0,0 +1,653 @@
+
+# Test configuration
+ENDPOINT_URL = "http://localhost:8000/receive"
+TEST_PHONE_NUMBER = "917567071072"
```
We should substitute these with env vars that we use across integration test cases:
```py
ENDPOINT_URL = os.getenv("RECIEVE_URL")
TEST_PHONE_NUMBER = os.getenv("PHONE_NUMBER_ID")
```

N1kunj1998 replied:
Added In Latest Commit.

mamuqsit commented on byoeb-v1/byoeb/tests/integration/test_early_return_already_onboarded.py:
```diff
@@ -0,0 +1,653 @@
+    return result
+
+
+async def test_onboarding_messages(language: str):
```
Need to add decorator here so the test is properly invoked for all languages:
```py
@pytest.mark.parametrize("language", [lang.value for lang in LanguageCode])
async def test_onboarding_messages(language: str):
    ...
```

N1kunj1998 replied:
Added In Latest Commit.

mamuqsit commented on byoeb-v1/byoeb/tests/integration/test_early_return_already_onboarded.py:
```diff
@@ -0,0 +1,653 @@
+        await asyncio.sleep(2)
+
+
+async def test_normal_question(language: str):
```
Need to add decorator here so the test is properly invoked for all languages:
```py
@pytest.mark.parametrize("language", [lang.value for lang in LanguageCode])
async def test_normal_question(language: str):
    ...
```

N1kunj1998 replied:
Added In Latest Commit.

mamuqsit commented on byoeb-v1/byoeb/tests/integration/test_early_return_already_onboarded.py:
```diff
@@ -0,0 +1,653 @@
+    print(f"🔄 Updating user language to {language}...")
+    success = await update_user_language(TEST_PHONE_NUMBER, language)
+    if not success:
+        print(f"⚠️  Warning: Could not update user language. Continuing anyway...")
```
`update_user_language` is not sufficient - we must register user using existing APIs instead. Even if the user does not exist, our tests must pass.
```diff
-    success = await update_user_language(TEST_PHONE_NUMBER, language)
-    if not success:
-        print(f"⚠️  Warning: Could not update user language. Continuing anyway...")
+    requests.delete(ENDPOINT_URL.replace("/receive", "/delete_users"), json=[TEST_PHONE_NUMBER]).raise_for_status()
+    requests.post(ENDPOINT_URL.replace("/receive", "/register_users"), json=[{
+        "phone_number_id": TEST_PHONE_NUMBER,
+        "user_location": {"district": "Jaipur"},
+        "user_type": "asha",
+        "user_language": language,
+        "user_name": "Test User",
+        "test_user": True
+    }]).raise_for_status()
```

Once this change is done, you can drop `get_mongo_collection()`, `update_user_language()`, and `update_user_language()`.

N1kunj1998 replied:
Added In Latest Commit.

mamuqsit commented on byoeb-v1/byoeb/tests/integration/test_early_return_already_onboarded.py:
```diff
@@ -0,0 +1,653 @@
+    print(f"🔄 Updating user language to {language}...")
+    success = await update_user_language(TEST_PHONE_NUMBER, language)
+    if not success:
+        print(f"⚠️  Warning: Could not update user language. Continuing anyway...")
```
See above comment (use /delete_users + /register_users)

N1kunj1998 replied:
Added In Latest Commit.

mamuqsit commented on byoeb-v1/byoeb/tests/integration/test_early_return_already_onboarded.py:
```diff
@@ -0,0 +1,653 @@
+    filter_dict = {"User.phone_number_id": phone_number}
+    user_doc = await collection.afetch_one(filter_dict)
+    
+    return user_doc
```
We can utilize existing API endpoint to perform this operation:
```py
async def find_user_by_phone(phone_number: str) -> Dict[str, Any]:
    response = requests.post(ENDPOINT_URL.replace("/receive", "/get_users"), json=[phone_number])
    response.raise_for_status()
    return response.json()[0]
```

N1kunj1998 replied:
Added In Latest Commit.

mamuqsit committed:
pydantic: make optional fields use explicit default=None (#104)

* pydantic: make optional fields use explicit default=None

According to https://docs.pydantic.dev/2.7/concepts/fields, the documented way to declare defaults is by using `default=`. This also solves VSC static analysis from underlining all such models where we expect defaults.
Furthermore:
> If you use typing.Optional, it doesn't mean that the field has a default value of None!

* message-response: make WhatsAppResponse.response_status optional in favor of test

mamuqsit committed:
kb-app: generate related questions in mr, te (#107)

* kb-app: generate related questions in mr, te as well

* /vector/search: expose related questions
This is needed to debug whether related questions are indeed being generated for all languages, and in the right script (Devanagari/Telugu)

mamuqsit committed:
has_interactive_list_additional_info: ensure list is not empty (#106)

mamuqsit committed:
bot-config: deduplicate answer_prompts, remove related_questions from prompt (#112)

mamuqsit committed:
translation: use gpt-4o-transcribe for speech-to-text (#99)

* refactor: rename async_azure_openai_whisper -> async_azure_openai_translator
This impl is generic for all OpenAI transcriber models, we can reuse it for GPT-4o Transcribe.

* mcp: support audio input in asha_chat

* dep-setup: suffix speech translators by their utility (_tts, _stt)

* app-config: allow specifying per-language stt and tts translation services

* app-config: use gpt-4o-transcribe for en, hi, te; whisper for mr
Goal was to use whisper for te as well, but whisper bails on processing te (unsupported)

* AzureOpenAIWhisperParamsEnum -> AzureOpenAISpeechParamsEnum

* dep-setup: make speech_* require user

* stt: gate gpt-4o-transcribe behind feature flag 'stt_latency_mitigation'

* tests: add integration tests for en, hi audio inputs

* azure-speech: fix failing test

* tests/integrations: add sample ogg files for audio messages

* AZURE_OPENAI_SPEECH_ENDPOINT: bail when undefined

* speech: reuse models having identical configuration

* speech_* selectors: accept lang, test_user instead of bulky User object for matching

* app-config: update speech model api versions 2024-06-01 -> 2024-10-21

* refactor: introduce a singular translator adapter for stt and tts

* fix failing test case
We shouldn't be using dependency_setup.py in our unit tests, but this is sufficient for time being

* dep-setup: read azure cognitive resource ids from env

mamuqsit committed:
mongodb: retire abstraction layers (#81)

* mongodb: first look at replacing abstractions with direct invocation

* mongodb: use mongomock library for unit tests

* byoeb-integrations: drop 'database' namespace

* repos: fix inconsistent typing

* ci/cd: use same staging-deploy environment as azure webapp (#83)

* ci-cd: temporarily enlist muqsit/docker-azure-env as staging branch

* Dockerfile: use python:3.11-slim-bullseye
This commit is expected to cause a deployment failure

* chat-app: overwrite default sqlite3 library with pysqlite3-binary
Expecting a successful deployment...

* Revert "ci-cd: temporarily enlist muqsit/docker-azure-env as staging branch"

This reverts commit 4fb32fb6fb353d4b6cad18737266f9d7421a02ad.

* Prompt modifications (#80)

* openapi: further documentation enhancements (#79)

* /get_users: use query params in place of request body
GET requests cannot have a request body according to HTTP specs - curl allows a GET with body anyway, but JavaScript's fetch() bails to process these requests.

* ashabot: model a PhoneNumberId type for auto validation

* apis/user: cleanup examples
These examples will be autogenerated by FastAPI based on pydantic models

* router: remove redundant 'response_class' field
These should be derived from the function return value itself

* apis/background_jobs: code cleanup

* apis/chat: code cleanup

* /get_users: convert to POST

* user: add integration tests for api endpoints

* scripts/onboarding.py: code cleanup

* tests: add integration test for onboarding.py

* apis: specify tags in router instead of in individual endpoints

* apis: code cleanup

* byoeb-core: add unit test for phone number validation

* byoeb-core: add coverage opts for pytest

* setup_scheduled_jobs: fail fast on exceptions
Any errors in job configuration will now cause deployment to fail instead of silently ignoring setup errors.

* scripts/onboarding.py: extract users only when needed

* message-db: delete unused delete_message_collection

* Prompt modifications (#80)

* openapi: further documentation enhancements (#79)

* /get_users: use query params in place of request body
GET requests cannot have a request body according to HTTP specs - curl allows a GET with body anyway, but JavaScript's fetch() bails to process these requests.

* ashabot: model a PhoneNumberId type for auto validation

* apis/user: cleanup examples
These examples will be autogenerated by FastAPI based on pydantic models

* router: remove redundant 'response_class' field
These should be derived from the function return value itself

* apis/background_jobs: code cleanup

* apis/chat: code cleanup

* /get_users: convert to POST

* user: add integration tests for api endpoints

* scripts/onboarding.py: code cleanup

* tests: add integration test for onboarding.py

* apis: specify tags in router instead of in individual endpoints

* apis: code cleanup

* byoeb-core: add unit test for phone number validation

* byoeb-core: add coverage opts for pytest

* setup_scheduled_jobs: fail fast on exceptions
Any errors in job configuration will now cause deployment to fail instead of silently ignoring setup errors.

* scripts/onboarding.py: extract users only when needed

* message-db: delete unused delete_message_collection

* repositories: move generic methods to a mongodb base repo class

* mongodb repo: find_all - use good default values

* mongodb base repo: return a (buffered) asynciterator instead of loading all records in memory

* message repo: return a buffered AsyncIterator instead of in-memory lists

* dyk: use same buffered fetch strategy

* ashabot: update cases to utilize new buffered fetch strategy

* repo: fix class hierarchy

* background-jobs: deprecate POST /schedule endpoint (#93)

* background-jobs: deprecate POST /schedule endpoint

* Return 200 OK with JSON "detail" as per standards
Addresses few points made by @a4i-architect

* POST /asha_logs: tweak setup, add integration tests (#92)

* POST /asha_logs: add integration tests
test_admin.py is expected to fail at two points

* admin.py: fix POST /asha_logs not returning expected response
This change is expected to fix the failing integration tests

* test_admin: use datetime.isoformat() in place of custom date formatter

* /asha_logs: stream logs server-side, process file client-side
Unlike XLSX, CSV is one of the better formats to use here as we can iteratively deliver data to user.

XLSX is a ZIP of XML files so it cannot be streamed.

* /get_bot_messages: add a 'limit=1000' parameter
We can leverage this to get a sample of messages for later /asha_logs integration tests

* test_admin: add test case to assert /asha_logs csv output format
This test will ensure the format contains few necessary attributes and that all records are sorted by incoming timestamp attribute.
/get_bot_messages is used to fetch a sample of up to a 100 messages.

* cleanup code, diff

* workflows: temporarily run cd
This change will be reverted after failing test is fixed

* Revert "workflows: temporarily run cd"

This reverts commit dbff763d4879516d839948bf142c17351613fa09.

* repositories: delete example usage file
This ought to be placed outside source, preferably in a docs/dev.md or the repo wiki

* repositories: cleanup comments
Docstrings already exist in base classes, so impls need not redefine them

* test-leaderboard: use mongomock-motor to clean up a plethora of monkeypatching

---------

Co-authored-by: Soumabha Ray Chaudhuri <Soumav@gmail.com>

N1kunj1998 committed:
corrected truncating text (#109)

N1kunj1998 committed:
Add unit tests to filter Test District from leaderboard before WhatsApp API calls (#111)

N1kunj1998 committed:
Addressed Review comments of using env instead of hardcoded values, and decorator for test invoking

a4i-architect reviewed (commented):
### Code Review Summary

---

#### 1. **Overall Objectives & Context**
- The changes aim to implement an early return mechanism for users already onboarded, preventing unnecessary LLM/vector store processing.
- Introduces robust language detection for response validation.
- Adds detailed debugging, validation, and reporting scripts.

---

#### 2. **File-specific Feedback & Issues**

**byoeb/chat_app/bot_config.json**
- Minor; adding `ALREADY_REGISTERED_DICT` for multi-language coverage is good.
- No issues.

**byoeb/constants/onboarding_text.py**
- Extensive addition of static messages for onboarding confirmation.
- Slight concern: multiple similar phrases can be localized or loaded via configuration for scalability.
- No logical issues.

**byoeb/services/channel/whatsapp.py**
- Debug print statements are good for tracing but should be replaced with proper logging or removed in production.
- No functional issues.

**byoeb/services/chat/message_handlers/user_flow_handlers/generate.py**
- **Main additions**:
  - Implemented `__detect_script()` to identify scripts (e.g., Telugu, Devanagari).
  - `__is_response_in_correct_language()` is critical but overly complex, with potential for misclassification and false negatives/positives.
  - **Potential issues**:
    - Script detection logic is basic and may misclassify mixed content or short responses.
    - Language checking is predominantly via script and keyword presence, which can be unreliable.
    - The fallback logic for responses in the wrong language mainly relies on `print()` for debugging, not exception handling or alerting.
    - When responses are detected as incorrect, fallback to translation is triggered without a clear boundary (may cause recursive issues if translation APIs fail).

- **Cyclomatic complexity concerns**:
  - `__is_response_in_correct_language()` is complex and could be simplified with clearer separation, e.g., separate language detection from content validation.
  - Logic branches for each language based on `detected_script` and keyword presence are somewhat convoluted.

- **Code duplication & type safety**:
  - String comparisons (`"te"`, `"devanagari"`, etc.) would benefit from Enums or constants.
  - Multiple `print()` statements; should migrate to structured logging.

**also in generate.py**:
- The fallback logic for responses in the user’s language vs. translation is essential but is currently run with print-based diagnostics.
- The logic validating whether the response is in the correct language could produce false negatives, leading to unnecessary re-translation.
- Using `detect_script` alone is fragile; consider integrating more robust language detection libraries (like `langdetect`, `cld3`).

**utils.py**:
- Minor; `is_onboard()` tweaks improve alignment with expected text matching.
- No critical issues.

**test_early_return_already_onboarded.py**
- **Positive**:
  - Well-structured, comprehensive integration testing.
  - Tests multiple languages and various message variants.
- **Concern**:
  - Heavy reliance on printing and sleep-based waiting (`asyncio.sleep()`, `time.sleep()`), which may cause flaky tests in CI.
  - The validation approach for responses relies on string containment, which may be brittle.
  - No explicit clean-up or setup ensures the user exists beforehand, only a `verify_user_exists()` check.

---

#### 3. **Design & Implementation Issues**

- **Language Detection**:
  - The current script detection method (`__detect_script()`) is brittle. It may misclassify responses with mixed scripts or short texts. Consider integrating more robust language detection libraries (`langdetect`, `cld3`).

- **Response Validation**:
  - Heavy reliance on keyword presence with fixed phrases is fragile.
  - For multi-language validation, better to use existing NLP language identification tools in tandem with script patterns.
  - Currently, the approach may produce false negatives, causing unnecessary re-translation or incorrect early returns.

- **Early Return Logic**:
  - The `__is_response_in_correct_language()` method infers correctness solely from script and keyword presence. It may misclassify legitimate responses, risking rejection or double translation.

- **Testing Strategy**:
  - Use of `asyncio.run()` in `__main__` is acceptable but consider extracting test logic further.
  - Tests could benefit from more explicit assert statements rather than relying solely on print logs.

- **Logging**:
  - Replace `print()` diagnostics with structured logging (`logger.debug()`), especially for production.

- **Potential Infinite Loops or Flaky Waiting**:
  - `wait_for_bot_response()` waits indefinitely; consider more robust timeout handling, correlation checks, or response content validation to avoid false positives/negatives.

---

#### 4. **Code Duplication & Maintainability**

- Multiple `response_source` validation blocks are similar across generate.py.
- Script/Language detection logic could be generalized into shared utility functions.
- Responses validation could be decoupled from message processing for cleaner separation of concerns.

---

#### 5. **Score (out of 100)**
- **Functionality**: 75/100
  - Achieves core goal of early detection for onboarded users.
- **Quality**:
  - Potential false negatives/positives due to simplistic language/script detection.
  - Overly complex conditional branches, especially in `__is_response_in_correct_language()`.
  - Extensive use of print statements for debugging, unsuitable for production.
  - Testing is comprehensive but could be more assertive.

**Final score: 78/100**

---

### Recommendations
- Simplify and/or replace `__detect_script()` with a more reliable language detection library.
- Normalize and clean `__is_response_in_correct_language()` to reduce false negatives.
- Use structured logging instead of prints.
- Add explicit assertions in tests to validate content.
- Consider a more scalable and modular approach for multi-language validation, e.g., language detection via external libraries before script validation.

---

**Summary**: The changes are functionally meaningful but need refinement in language detection robustness, validation logic clarity, and testing assertions for more reliable integration.

N1kunj1998 committed:
replace find_user_by_phone with /get_users, using /delete_users + /register_users endpoints, removed get_mongo_collection and update_user_language

a4i-architect reviewed (commented):
The proposed code modifications are extensive, primarily focused on improving language detection, robustly handling onboarding responses, and adding a dedicated test suite for early response validation. Here's an in-depth review:

---

### Strengths & Good Practices
- **Explicit Language Detection**: The `__detect_script` method with Unicode range checks enhances robustness in language validation, especially for Telugu (`\u0C00-\u0C7F`) and Devanagari scripts.
- **Early Return for Onboarding**: The `is_onboard` function now normalizes whitespace and compares phrases in a case-insensitive manner (for English) or with careful substring matching, reducing false negatives.
- **Test Suite**: The comprehensive integration test scripts (`test_early_return_already_onboarded.py`) correctly simulate user onboarding and normal question flows across multiple languages with thorough debug logs and reporting.
- **Logging & Debugging**: Extensive print statements aid debugging, especially in asynchronous waits and response validations.
- **Validation & Reporting**: The validation validation effectively checks for the presence of "already registered" indicators, offering summarized and detailed reports.

---

### Issues & Improvements

#### 1. **Overcomplicated Language Detection Logic**
- **Duplication & Overuse of normalize_whitespace**:
  - The detection logic compares normalized strings with substring checks, which could lead to false positives/negatives. 
  - *Suggestion*: Use a consistent, well-defined string comparison, perhaps leveraging regex or keyword matching with word boundaries to improve accuracy.
- **Handling of Non-Standard Text**:
  - The use of `detected_script` is good but may miss mixed scripts or transliterations.
  - *Suggestion*: Enhance detection with NLP tools or language models for better accuracy, or leverage existing language detection libraries (like `langdetect` or `fasttext`).

#### 2. **Redundant or Ill-Typed Code**
- **`__is_response_in_correct_language`**:
  - Contains duplicated logic, some of which can be streamlined with a mapping of scripts to languages.
  - Heavy reliance on substring presence for language confirmation is brittle.
- **`is_onboard`**:
  - The current pattern of lowering, replacing hyphens, and checking substrings is acceptable but could be replaced by a more robust keyword list or regex matching for increased reliability.
- **`try-except` blocks** around cache access are proper, but ensure cache methods themselves are correctly typed.
- **`wait_for_bot_response`**:
  - It correctly filters messages, but the logic for handling message categories can be simplified by defining a set of expected categories and checking membership.
  - Consider adding more explicit logging for unexpected message structures.

#### 3. **Design & Architecture**
- **Testing vs. Production**:
  - The test script is integration-heavy and depends on an accessible server. It is logically sound but could benefit from mocks or fixtures for unit tests.
- **Handling of `response_source`**:
  - The logic for falling back to translation if the language detection fails is practical, but the translation function may introduce delay.
  - *Suggestion*: Cache translations, or perform language detection before translation to minimize overhead.
  
#### 4. **Miscellaneous**
- **Response Validation**:
  - The validation code searches for indicators of "already registered" messages, which is good. But matchings could be improved with regex or NLP approaches.
- **Code Duplication**:
  - The test suite has repeated code blocks for registration, messaging, and response handling which could be encapsulated.
- **Code Modularization**:
  - The large test script would benefit from splitting into smaller functions or classes, especially for registration, message sending, waiting, and validation.
  - APIs for message sending and response validation could be decoupled into helper modules.

---

### Scoring: **75/100**

- **Pros**:
  - Thorough language detection improvements
  - Well-structured testing framework
  - Clear debug and validation approach

- **Cons**:
  - Added complexity in language detection and validation logic (can be simplified)
  - Some code duplication and lack of modularity
  - Slightly brittle string/substring matching approach
  - Tests rely on external server availability without mocks

---

### Recommendations:
- **Simplify language detection** with a dedicated NLP or language detection library.
- **Improve matching logic** in `is_onboard` with regex or token-based matching.
- **Encapsulate repetitive code** for registration, messaging, and validation.
- **Add unit tests** (with mocks) for key functions like language detection and response validation.
- **Ensure cache and API calls** handle exceptions gracefully and are well-typed.

With these enhancements, the code would be more maintainable, reliable, and clearer.

**Final Verdict:** **Accept with Minor to Moderate Refactoring (Score: 75/100)**

a4i-architect reviewed (commented):
### Summary of Findings

#### 1. **Overall Code Quality & Consistency**
- The codebase is undergoing substantial refactoring, especially around language detection (`__detect_script`, `__is_response_in_correct_language`) and onboarding message handling.
- Multiple debugging `print` statements and verbose logs are included, which should be set to a debug mode or removed for production.

#### 2. **Major Issues & Concerns**

**a. Unnecessary and Redundant Code**
- The `__detect_script` and `__is_response_in_correct_language` are duplicated and overly complex, with some functions not being fully optimized or tested.
- The function `__is_response_in_correct_language()` has an overly complicated logic for language detection, which might lead to false positives/negatives. Language detection via script ranges is approximate; consider integrating a dedicated language detection library (e.g., `langdetect`, `fastText`).

**b. Handling of Language Detection and Response Validation**
- The approach to confirm response correctness based on script presence and keyword hints (like "है", "आहे", "నమోదు చేయబడ్డారు") is brittle.
- Marathi and Hindi responses may both be in Devanagari, so relying heavily on script detection and keywords might misclassify.

**c. Code Duplication & Complexity**
- The `handle_message_generate_workflow()` is quite lengthy; the logic around onboarding, fallback translation, and response validation could be refactored into smaller, reusable methods.
- The way `response_source` is validated seems over-complicated; some code branches are redundant or commented-out.

**d. Testing & Validation**
- The `test_early_return_already_onboarded.py` is comprehensive but is:

  - **Too verbose**: Many print statements, which should switch to proper logging.
  - **Dependent on external requests**: Make sure to mock API responses for reliable testing.
  - **Handling various edge cases**: For example, multiple categories, missing fields, or unorthodox message structures.

---

### Specific Issue Highlights

- **Language Detection (core logic)**:
  - `__detect_script()` only detects based on Unicode ranges; doesn't handle mixed scripts or non-standard cases.
  - `__is_response_in_correct_language()` makes assumptions about keywords, which may not hold in real-world data, especially for Marathi vs Hindi.

- **Response Validation Risks**:
  - Relying on substring matching in long responses is fragile; consider NLP-based language identification.
  - The use of `any()` with keyword lists is quick but not robust for production. Use language detection libraries.

- **Error Handling & Fallbacks**
  - `try-except` blocks around cache updates are okay but should log exceptions meaningfully.
  - When response is in the wrong language, fallback to translation is good; ensure translated response matches expected language.

- **Test Script Concerns**
  - The script mixes test setup, execution, and validation into one file. Consider dividing it into setup, tests, and validation modules.
  - The validation logic depends on string matching with localized phrases, which might have inaccuracies.

---

### Design & Complexity Evaluation

- The language validation functions, while ambitious, are overly intricate for what can be achieved more reliably with dedicated NLP tools.
- The message handling employs deep nested data structures and multiple levels of checks; reducing complexity via helper functions or classes would improve maintainability.
- The response validation based on scripts and keywords introduces potential false negatives, especially for Marathi vs Hindi, which both use Devanagari.

---

### Recommended Actions

- **Simplify language detection**: Use NLP libraries (`langdetect`, `langid`) instead of Unicode ranges and keyword lists.
- **Refactor large methods**: Break down `__create_user_message()` and `handle_message_generate_workflow()` into smaller, focused functions.
- **Remove debug statements**: Use a proper logging framework with debug levels.
- **Enhance testing**: Implement mock responses for API calls; verify language detection accuracy.
- **Avoid repeating logic**: For example, response validation patterns should be centralized.

### Score: **75/100**

Justification:
- Good comprehensive testing and validation logic.
- Significant scope for refactoring, simplification, and robustness.
- Barring the complexity, the core ideas are correct and aligned with best practices.

**Accept with suggestions for improvements**:
- Clean up verbose logs.
- Use NLP for language detection.
- Modularize large functions.
- Improve error handling.

---

**References:**
- [Effective Language Detection Techniques](https://stackify.com/language-detection-in-python/)
- [Refactoring for Maintainability](https://refactoring.guru/)
- [Best practices for API testing](https://rest-assured.io/)

---

If you want a detailed line-by-line review or specific refactoring suggestions, let me know!

N1kunj1998 committed:
Merge branch 'a4i/staging' into nikunj/a4i/staging-proactive-return-already-onboarded

a4i-architect reviewed (commented):
### Review Summary

#### General Observations:
- Significant enhancements for language detection and validation logic, which is excellent.
- The "early return" logic for already onboarded users is well-integrated into the message workflow.
- Extensive debug/print statements aid debugging but should be conditioned or removed in production.
- The test script is comprehensive, covering onboarding in multiple languages and normal flow testing.

---

### 1. Code Quality & Duplication
- **Redundant code:** Many print statements (debug logs) are directly embedded; consider using a logging framework with log levels.
- **Repeated code:** Similar language validation logic appears in multiple places—could be refactored into a dedicated utility function.
- **Test script complexity:** The monolithic pytest script is lengthy; it could benefit from refactoring into smaller classes/functions for clarity.

### 2. Language Detection & Validation
- **`__detect_script` method:** Good, but could be extended for Hindi/Marathi detection (e.g., if response contains certain script-specific Unicode or words). Currently, we check only for Telugu, Hindi, and Latin.
- **`__is_response_in_correct_language`:** 
  - The logic is complex and attempts to heuristically confirm language correctness. While necessary, it should be more robust
  - For Marathi: relying on presence of "आहे" etc. can be brittle; consider more comprehensive language detection libraries or pattern matching.
  - Current heuristic (word presence) may false positive/negative; consider integrating with established libraries like `langdetect` or `cld3`.
- **Potential issue:** If response contains mixed-language content, detection may fail.

### 3. Code Hygiene & TypeSafety
- Type annotations are present but inconsistent—add types for private methods consistently.
- Exception handling around cache/database calls is good but could be more centralized.
- The `utils.is_onboard()` function is adjusted for multi-language support but the whitespace normalization and phrase matching logic is somewhat brittle; more formal NLP-based language detection could improve robustness.
- The `__detect_script()` function assumes specific Unicode ranges; expand coverage if needed.

### 4. Performance & Complexity
- Language validation heuristics add complexity and may slow down overall response processing.
- The sleep intervals in response waiting are acceptable but could be parameterized or based on server response times.

### 5. Unnecessary & Duplicate Code
- Multiple debug print checks for API responses and message categories could be turned into proper logging.
- The validation report generation repeats some logic from test functions—consider centralizing validation reports.
- The code handling onboarding and normal message processing can be modularized into reusable components.

---

### Recommendations:

- **Implement a proper logging system** instead of raw print statements.
- **Refactor language detection/validation** into a dedicated module with more robust methods possibly using `langdetect` or `cld3`.
- **Consolidate onboarding message detection** into a single utility with clear normalization.
- **Simplify `__is_response_in_correct_language`**: focus on Unicode script detection, and, if possible, integrate NLP language detection libraries.
- **Limit debug verbosity**: switch to debug level logs with a configuration switch.
- **Test code**: Break into smaller functions/classes; reduce verbosity for clarity.

### Score: 75/100
- **Positives:** Strong intent, improving multilingual handling, and user experience.
- **Areas for improvement:** Code modularity, error handling, and robustness of language heuristics.

**Overall, accept with minor improvements**, explicitly addressing logging, modularity, and language detection robustness.

a4i-architect reviewed (commented):
### Code Review Summary

---

### 1. **byoeb-v1/byoeb/byoeb/chat_app/bot_config.json**
- **No issues**; updates are mainly JSON content with detailed rules for responses and translations.

### 2. **byoeb-v1/byoeb/byoeb/constants/onboarding_text.py**
- **Good addition** for static onboarding messages.
- **Minor note**: The list of messages for each language could be compressed or made more systematic, but it's not critical.

### 3. **byoeb-v1/byoeb/byoeb/services/channel/whatsapp.py**
- **Use of print debugging statements** is excessive; should be replaced with proper logging.
- **No functional issues** identified; the conditional structure for message type is clear.
- **Suggestion**: Consider unifying or centralizing message preparation logic for better maintenance.

### 4. **byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py**
- **Major addition**: Language script detection and validation functions (`__detect_script`, `__is_response_in_correct_language`).
  - The script detection logic (using Unicode ranges) is solid.
  - Language-specific validation based on script + indicator words is reasonable but fragile; e.g., relying on specific words (like "आहे" in Marathi) may be brittle.
  - **Complex cyclomatic code**: The multi-branch logic for language validation could be simplified with a mapping approach. It’s currently verbose.
- **Handling of fallback responses and language validation**: effective for preventing language mismatches, but could benefit from a cleaner, more configurable approach.
- **Additional improvements**:
  - The `__is_response_in_correct_language` could be optimized with constants for word indicators.
  - Mixed use of `try-except` for cache operations: acceptable, but may hide bugs if exceptions are broad.
- **Logging**: Uses `print()` for debugging; should be replaced with a logging framework.

### 5. **byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/process.py**
- **Significant addition**:
  - Early return for onboarding messages: well-intentioned but a bit convoluted.
  - Using `utils.is_onboard()` to identify onboarding: good, but logic could be clearer with explicit checks rather than nested conditionals.
  - When onboarded, the code fetches messages from `ALREADY_REGISTERED_DICT` for response; the translation process from English to user's language appears to be duplicated.
- **Potential issues**:
  - The fallback to default `message_english_text` for non-onboarding is appropriate.
  - **Complex control flow**: The block handling onboarding and fallback could be streamlined; e.g., separate functions.
  - Caching is handled with `try-except`, which is good.
- **Category setting logic**: 
  - Defaulting to `INTERACTIVE_LIST` for text responses might be okay but could be made explicit.
  - Overly broad assumptions about message types when setting `message_type`.
- **Overall**: Careful refactoring for clarity and modularity is recommended.

### 6. **byoebeb/utils/utils.py**
- **Minor fix**: The check for `constants.ROW_TEXTS` now ensures it’s not empty, which is good.

---

### **General Issues & Recommendations**

- **Excessive use of print()**: Replace with a structured logging system (`logging` module).
- **Language validation complexity**:
  - The approach using Unicode blocks and indicator words is fragile.
  - Consider external language detection libraries or more robust heuristics.
  - Maintain a configuration-driven approach rather than hardcoded words.
- **Duplication & Complexity**:
  - The onboarding logic and response handling are somewhat duplicated across files. Extract shared functions for reusability.
- **Exception handling**:
  - Broad `try-except` blocks with silent failures may hide bugs.
  - Log exceptions explicitly, or handle specific exception types.
- **Maintainability**:
  - Split large conditionals (`if-else`) into smaller, well-named functions.
- **Test fixtures**:
  - The integration test script is comprehensive, but verbose. Consider parameterizing more parts.
  - Some assertions are missing; adding checks on response content would strengthen validation.

---

### **Score: 78/100**

**Rationale**:
- The code addresses the requirements and introduces valuable language validation features.
- However, it suffers from verbosity, debugging clutter, and fragile heuristics.
- To reach 85+ and be production-ready:
  - Replace debug prints with logs.
  - Simplify language detection with a configurable, external library.
  - Modularize onboarding handling.
  - Add assertions or explicit checks in tests.
  - Improve exception granularity and error handling.

---

### **Conclusion**
The changes are promising but need refinement for robustness and clarity. Minor issues can be addressed with refactoring and better testing. Overall, **acceptable with recommended improvements**.

mamuqsit commented on byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py:
```diff
@@ -301,6 +301,96 @@ async def __create_source_audio(
+            has_marathi_indicators = any(ind in response_source for ind in marathi_indicators)
+            has_hindi_indicators = any(ind in response_source for ind in hindi_indicators)
+            
+            # Count occurrences for better detection
+            marathi_count = sum(response_source.count(ind) for ind in marathi_indicators)
+            hindi_count = sum(response_source.count(ind) for ind in hindi_indicators)
+            
+            # If it has more Hindi indicators than Marathi, it's likely Hindi
+            if hindi_count > marathi_count and hindi_count > 2:
+                print(f"[__is_response_in_correct_language] Marathi user but response has more Hindi indicators ({hindi_count}) than Marathi ({marathi_count})")
+                return False
+            
+            # If it has Hindi indicators but no Marathi indicators, it's likely Hindi
+            if has_hindi_indicators and not has_marathi_indicators and hindi_count > 1:
+                print(f"[__is_response_in_correct_language] Marathi user but response has Hindi indicators ({hindi_count}) without Marathi indicators")
+                return False
+            
+            # If it has Marathi indicators, it's likely correct
+            if has_marathi_indicators:
+                return True
+            
+            # If no clear indicators but it's in Devanagari, be lenient (might be correct but generic)
+            # Only reject if we're very confident it's Hindi
+            if hindi_count > 3:
+                print(f"[__is_response_in_correct_language] Marathi response has many Hindi indicators ({hindi_count}) but no Marathi indicators")
+                return False
+            
+            # Default to accepting if in Devanagari (might be correct Marathi)
+            return True
+        
+        # For Hindi, must be in Devanagari script
+        elif user_language == "hi":
+            return detected_script == "devanagari"
+        
+        # For English, must be in Latin script
+        elif user_language == "en":
+            return detected_script == "latin" or detected_script == "unknown"
+        
+        # Unknown language, accept it
+        return True
```
These utility functions should be moved to `byoeb/utils/utils.py`, we can benefit from reusing these functions in tests where we test for Devanagari script (e.g., in byoeb/tests/integration/utils/test_embedding_cache.py)

mamuqsit commented on byoeb-v1/byoeb/byoeb/chat_app/bot_config.json:
```diff
@@ -28,10 +28,10 @@
-                    "en": "3. Set response_src = response_en.",
-                    "mr": "3. Translate response_en into spoken Marathi to generate response_src, following the below rules:\n\n3a. Use simple Marathi that is spoken in everyday life. Do not use complex or uncommon Marathi words. Do not use words that are typically used only in formal or written Marathi. Also, feel free to use English words that are commonly used in Marathi, like \"लिवर\", \"इंजे ...
-                    "te": "3. Translate response_en into spoken Telugu to generate response_src, following the below rules:  \n\n<rules>  \n\n1. Use simple Telugu that is spoken in everyday life. Do not use complex or uncommon Telugu words. Do not use words that are typically used only in formal or written Telugu. Also, feel free to use English words that are commonly used in Telugu, like \"లివర్ ...
+                    "hi": "3. Translate response_en and related_questions_en into spoken Hindi to generate response_src and related_questions_src, following the below rules:\n\n3a. Use simple Hindi that is spoken in everyday life. Do not use complex or uncommon Hindi words. Do not use words that are typically used only in formal or written Hindi. Also, feel free to use English words that are comm ...
+                    "en": "3. Set response_src = response_en. Set related_questions_src = related_questions_en. ",
+                    "mr": "3. Translate response_en and related_questions_en into spoken Marathi to generate response_src and related_questions_src, following the below rules:\n\n3a. Use simple Marathi that is spoken in everyday life. Do not use complex or uncommon Marathi words. Do not use words that are typically used only in formal or written Marathi. Also, feel free to use English words that  ...
+                    "te": "3. Translate response_en and related_questions_en into spoken Telugu to generate response_src and related_questions_src, following the below rules:  \n\n3a. Use simple Telugu that is spoken in everyday life. Do not use complex or uncommon Telugu words. Do not use words that are typically used only in formal or written Telugu. Also, feel free to use English words that ar ...
```
These changes appear to revert past modifications from https://github.com/A4i-tech/byoeb/commit/e0079290d45fa1b12bcbed7f682a236196a03994.

`git fetch && git merge origin/a4i/staging` should address these.

mamuqsit commented on byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py:
```diff
@@ -383,18 +492,26 @@ async def __create_user_message(
                 else:
                     cache[user_language] = media_info
-                self.embedding_cache.update(cache_id, cache)
+                try:
+                    self.embedding_cache.update(cache_id, cache)
+                except Exception as e:
+                    print(f"Warning: Embedding cache update failed: {e}. Continuing without cache update.")
```
Failsafe here can be omitted, EmbeddingCache.update is supposed to handle its own side-effects.

mamuqsit commented on byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/generate.py:
```diff
@@ -712,8 +909,12 @@ async def handle_message_generate_workflow(
                     miss_thresh = cache_result[0] if cache_result is not None else None
-                    cache_result = self.embedding_cache.store(embedding, cache_val)
-                    cache_result = miss_thresh, *cache_result[1:]
+                    try:
+                        cache_result = self.embedding_cache.store(embedding, cache_val)
+                        cache_result = miss_thresh, *cache_result[1:]
+                    except Exception as e:
+                        print(f"Warning: Embedding cache store failed: {e}. Continuing without cache.")
+                        cache_result = None, None, None
```
Failsafe here can be omitted, EmbeddingCache.store is supposed to handle its own side-effects.

mamuqsit commented on byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/process.py:
```diff
@@ -142,8 +142,44 @@ async def handle_process_message_workflow(
+            source_text = message.message_context.message_source_text
+            query_en = source_text
+            query_en_addcontext = source_text
+            query_type = "asha_work_related"
```
This can be replaced with `query_type = self._asha_work_related`

mamuqsit commented on byoeb-v1/byoeb/byoeb/utils/utils.py:
```diff
@@ -61,13 +62,57 @@ def is_onboard(
     lang: str = LanguageCode.ENGLISH.value
 ):
+    # Safety check: if text is None or empty, return False
+    if not text or not isinstance(text, str):
+        print(f"[is_onboard] Invalid text input: {text} (type: {type(text)})")
+        return False
```
`text` is expected to be str here based on parameter typing, so this check can be relaxed.
